File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // target ,  nums array as input
        // find two val that add up to target
        // return their indices
        // x + y = target
        // use value from nums as x, then the equation will become something like 2 + y = 9 
        // y = 9 - 2;
        // y = target - nums[i];
        // O(1) == set and map
        // if i use set == > keys (individual values) but for this problem i need to return the indices...
        // hashmap ==>  nums[i] , i


        HashMap<Integer,Integer> freqMap = new HashMap<>();

        for ( int i = 0; i < nums.length; i++ ){
            int y = target - nums[i];
            if(freqMap.containsKey(y)){
                return new int[] {freqMap.get(y),i};
            }

            freqMap.put(nums[i] , i);
        }

        return new int[] {};
    }
}

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hash_map = {}
        # x + y = target
        # y = target - x
        for index, num in enumerate(nums):
            y = target - num
            if y in hash_map:
                return [hash_map[y] , index]
            hash_map[num] = index
        
        return []


        